### ISSUE TITLE
How to visualize a category or an area in an S3DIS room

### ISSUE THREAD

**Issue by xiaoyangyang123** [AnTao97/dgcnn.pytorch#79]
Hello, I would like to ask, how to visualize a certain category in S3DIS, following your steps can only show the visualization of the whole room. But if I want to show a visualization of a certain category of a room, how can I do it? Just like the visualization shown in your steps.

**Comment by antao97** [comment]
Sorry, I do not get your point. Do you want to visualize a whole room but a specific class is colored?

**Comment by xiaoyangyang123** [comment]
Yes, for example the category chairs in office 1 in area 5.

**Comment by antao97** [comment]
Replace 
https://github.com/antao97/dgcnn.pytorch/blob/97785863ff7a82da8e2abe0945cf8d12c9cc6c18/main_semseg.py#L106-L107

with the following code:
```
visual_class = [8]
if pred[i][j] in visual_class:
    RGB.append(semseg_colors[int(pred[i][j])])
else:
    RGB.append([255, 255, 255])
if seg[i][j] in visual_class:
    RGB_gt.append(semseg_colors[int(seg[i][j])])
else:
    RGB_gt.append([255, 255, 255])
```

**Comment by xiaoyangyang123** [comment]
Thank you, I'll give it a try.

**Comment by xiaoyangyang123** [comment]
你好，我按照您的代码以及您的预训练模型可视化出来的效果如图，我看到您的椅子是只有红色的，为什么我的椅子以及整个点云图是七彩斑斓的。

<img width="107" alt="椅子" src="https://user-images.githubusercontent.com/87469804/198495240-c182d999-8740-4d81-a06c-ebbd75548c08.png">
<img width="688" alt="全景" src="https://user-images.githubusercontent.com/87469804/198495249-ca0985da-239e-4723-a3fd-2672bc5ae3a6.png">

**Comment by antao97** [comment]
你再检查一下？应该是颜色和点云的匹配不太对，我这边没有问题。正常情况下，如果预测的结果很差也会是整片颜色预测错误，不会出现这种七彩斑斓的情况。

**Comment by xiaoyangyang123** [comment]
好的，我再尝试一下，可能是我将一下代码行的缩进进行了更改导致的吗？但是如果不这样处理的话，我这边机子会报错。
`def test(args, io):
    all_true_cls = []
    all_pred_cls = []
    all_true_seg = []
    all_pred_seg = []
    for test_area in range(1, 7):
        visual_file_index = 0
        test_area = str(test_area)
        if os.path.exists("data/indoor3d_sem_seg_hdf5_data_test/room_filelist.txt"):
            with open("data/indoor3d_sem_seg_hdf5_data_test/room_filelist.txt") as f:
                for line in f:
                    if (line[5]) == test_area:
                        break
                    visual_file_index = visual_file_index + 1
        if (args.test_area == 'all') or (test_area == args.test_area):
            test_loader = DataLoader(S3DIS(partition='test', num_points=args.num_points, test_area=test_area),
                                     batch_size=args.test_batch_size, shuffle=False, drop_last=False)

            device = torch.device("cuda" if args.cuda else "cpu")

            # Try to load models
            semseg_colors = test_loader.dataset.semseg_colors
            if args.model == 'dgcnn':
                model = DGCNN_semseg().to(device)
            else:
                raise Exception("Not implemented")

            model = nn.DataParallel(model)
            model.load_state_dict(torch.load(os.path.join(args.model_root, 'model_%s.t7' % test_area)))
            model = model.eval()
            test_acc = 0.0
            count = 0.0
            test_true_cls = []
            test_pred_cls = []
            test_true_seg = []
            test_pred_seg = []
            with torch.no_grad():
                for data, seg in test_loader:
                    data, seg = data.to(device), seg.to(device)
                    data = data.permute(0, 2, 1)
                    batch_size = data.size()[0]
                    seg_pred = model(data)
                    seg_pred = seg_pred.permute(0, 2, 1).contiguous()
                    pred = seg_pred.max(dim=2)[1]
                    seg_np = seg.cpu().numpy()
                    pred_np = pred.detach().cpu().numpy()
                    test_true_cls.append(seg_np.reshape(-1))
                    test_pred_cls.append(pred_np.reshape(-1))
                    test_true_seg.append(seg_np)
                    test_pred_seg.append(pred_np)
                    # visiualization
                    visualization(args.visu, args.visu_format, args.test_area, data, seg, pred, visual_file_index,
                                  semseg_colors)
                    visual_file_index = visual_file_index + data.shape[0]
                if visual_warning and args.visu != '':
                    print('Visualization Failed: You can only choose a room to visualize within the scope of the test area')
                test_true_cls = np.concatenate(test_true_cls)
                test_pred_cls = np.concatenate(test_pred_cls)
                test_acc = metrics.accuracy_score(test_true_cls, test_pred_cls)
                avg_per_class_acc = metrics.balanced_accuracy_score(test_true_cls, test_pred_cls)
                test_true_seg = np.concatenate(test_true_seg, axis=0)
                test_pred_seg = np.concatenate(test_pred_seg, axis=0)
                test_ious = calculate_sem_IoU(test_pred_seg, test_true_seg)
                outstr = 'Test :: test area: %s, test acc: %.6f, test avg acc: %.6f, test iou: %.6f' % (test_area,
                                                                                                        test_acc,
                                                                                                        avg_per_class_acc,
                                                                                                        np.mean(
                                                                                                            test_ious))
                io.cprint(outstr)
            all_true_cls.append(test_true_cls)
            all_pred_cls.append(test_pred_cls)
            all_true_seg.append(test_true_seg)
            all_pred_seg.append(test_pred_seg)

    if args.test_area == 'all':
        all_true_cls = np.concatenate(all_true_cls)
        all_pred_cls = np.concatenate(all_pred_cls)
        all_acc = metrics.accuracy_score(all_true_cls, all_pred_cls)
        avg_per_class_acc = metrics.balanced_accuracy_score(all_true_cls, all_pred_cls)
        all_true_seg = np.concatenate(all_true_seg, axis=0)
        all_pred_seg = np.concatenate(all_pred_seg, axis=0)
        all_ious = calculate_sem_IoU(all_pred_seg, all_true_seg)
        outstr = 'Overall Test :: test acc: %.6f, test avg acc: %.6f, test iou: %.6f' % (all_acc,
                                                                                            avg_per_class_acc,
                                                                                            np.mean(all_ious))
        io.cprint(outstr)
        print(all_ious)
`

**Comment by xiaoyangyang123** [comment]
我更改了这几行的缩进 
       all_true_cls.append(test_true_cls)
        all_pred_cls.append(test_pred_cls)
        all_true_seg.append(test_true_seg)
        all_pred_seg.append(test_pred_seg)

**Comment by xiaoyangyang123** [comment]
好的，是因为我改了main的缩进导致的，谢谢您。
